File: src/test/java/de/pflugradts/passbird/application/util/SystemOperationFaker.java
package de.pflugradts.passbird.application.util;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.Map;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.BDDMockito.given;
import static org.mockito.Mockito.lenient;
import static org.mockito.Mockito.mock;

@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
public class SystemOperationFaker {

    private SystemOperation systemOperation;
    private final Map<String, Path> paths = new HashMap<>();

    public static SystemOperationFaker faker() {
        return new SystemOperationFaker();
    }

    public SystemOperationFaker fakeSystemOperation() throws IOException {
        this.systemOperation = mock(SystemOperation.class);
        lenient().when(systemOperation.newInputStream(any())).thenReturn(mock(InputStream.class));
        lenient().when(systemOperation.newOutputStream(any())).thenReturn(mock(OutputStream.class));
        return this;
    }

    public SystemOperationFaker forInstance(final SystemOperation systemOperation) {
        this.systemOperation = systemOperation;
        return this;
    }

    public SystemOperationFaker withPath(final String uri, final Path path) {
        this.paths.put(uri, path);
        return this;
    }

    public SystemOperationFaker withConsoleDisabled() {
        given(systemOperation.isConsoleAvailable()).willReturn(false);
        return this;
    }

    public SystemOperationFaker withPasswordFromConsole(char[] password) {
        given(systemOperation.isConsoleAvailable()).willReturn(true);
        given(systemOperation.readPasswordFromConsole()).willReturn(password);
        return this;
    }

    public SystemOperation fake() {
        paths.forEach((uri, path) -> given(systemOperation.getPath(uri)).willReturn(path));
        return systemOperation;
    }

}

File: src/test/kotlin/de/pflugradts/passbird/application/util/SystemOperationFaker.kt
package de.pflugradts.passbird.application.util

import de.pflugradts.passbird.application.Directory
import de.pflugradts.passbird.application.FileName
import de.pflugradts.passbird.application.toDirectory
import io.mockk.every
import io.mockk.mockk
import java.io.IOException
import java.io.OutputStream
import java.nio.file.Path
import java.nio.file.Paths
import java.time.Clock

fun fakeSystemOperation(
    instance: SystemOperation,
    withClock: Clock = Clock.systemUTC(),
    withPaths: List<Pair<String, Path>> = emptyList(),
    withDirectoryResolvingToFileName: Triple<Directory, FileName, Path>? = null,
    withIoException: Boolean = false,
) {
    every { instance.clock } returns withClock
    every { instance.newInputStream(any()) } returns mockk()
    every { instance.newOutputStream(any()) } returns mockk()
    every { instance.writeToSensitiveFile(any(), any()) } answers {
        secondArg<(OutputStream) -> Unit>().invoke(mockk<OutputStream>(relaxed = true))
        Paths.get("")
    }
    withPaths.forEach { every { instance.getPath(it.first.toDirectory()) } returns it.second }
    withDirectoryResolvingToFileName?.run { every { instance.resolvePath(first, second) } returns third }
    every { instance.exit(any()) } returns Unit
    if (withIoException) every { instance.resolvePath(any(Directory::class), any(FileName::class)) } throws IOException()
    if (withIoException) every { instance.resolvePath(any(Directory::class), any(Directory::class)) } throws IOException()
    if (withIoException) every { instance.writeToSensitiveFile(any(), any()) } throws IOException()
}
